### PR TITLE
Cleanup with pylint

### PR DESCRIPTION
**Pull request by svaningelgem**
- Cleanup the pylint things that were ignored. Most notably the cyclic-imports.
- Made all internal references relative. Don't import pysparkling itself anymore from within the library.
- Removed (object) inheritance. It's the only way anyway on Python3.
- `Tokenizer::next` failed on pylint. Likely because pylint confuses it with the internal `next` command? Renamed it to get_next to silence the error... I don't like it, but I prefer to have the switch enabled to capture other use cases.

### CODE REVIEW

File: pysparkling/context.py [hunk 1]
```diff
@@ -194,6 +193,9 @@ def parallelize(self, x, numSlices=None):
         :rtype: RDD
         """
         if numSlices is None or numSlices <= 1:
+            # pylint: disable=import-outside-toplevel, cyclic-import
```
Comment (svenkreiss) [inline pysparkling/context.py:196]: This was importable at the top level before. Can you elaborate why that cannot be done anymore?

File: pysparkling/rdd.py [hunk 2]
```diff
@@ -167,9 +167,9 @@ def seqFuncByKey(tc, i):
                 r[k] = seqFunc(r[k], v)
             return r
 
-        def combFuncByKey(l):
+        def combFuncByKey(lst):
```
Comment (svenkreiss) [inline pysparkling/rdd.py:170]: To disentangle these PRs and to move forward, I'd suggest not to mix issues and not convert `l` to `lst`. I am against abbreviations in variable names and for some of your replacements `l` is not actually a list so we would have to have a discussion about those names which seems to be out of scope for this PR.

**Review by svenkreiss — CHANGES_REQUESTED** [review]
Thanks for all this work. Some comments below.

File: pysparkling/__version__.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+from ._version import get_versions
```
[the remark this answers is not in the document]
Comment (svaningelgem) [reply pysparkling/__version__.py:1]: Yes indeed. I couldn't fix it in another way because it would recursively call the library and as the first step the version again.

File: pysparkling/sql/group.py [hunk 4]
```diff
@@ -1,10 +1,9 @@
-from pysparkling.sql.column import Column
-from pysparkling.sql.dataframe import DataFrame
-# pylint: disable=W0622
-from pysparkling.sql.functions import avg, count, lit, max, mean, min, parse, sum
+from .column import Column
+# pylint: disable=redefined-builtin
+from .functions import avg, count, lit, max, mean, min, parse, sum
```
[the remark this answers is not in the document]
Comment (svaningelgem) [reply pysparkling/sql/group.py:3]: I didn't want to change too much, but I'll fix this.
What do you think about `from . import functions as F`?

File: pysparkling/sql/internal_utils/readers/common.py [hunk 5]
```diff
@@ -1,8 +1,8 @@
-from pysparkling.sql.internal_utils.readers.csvreader import CSVReader
```
[the remark this answers is not in the document]
Comment (svaningelgem) [reply pysparkling/sql/internal_utils/readers/common.py:1]: Same as before, I did not want to cleanup anything here. Just make everything relative.
But I adjust as you said. It's cleaner that way.

File: pysparkling/sql/group.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (svenkreiss) [reply pysparkling/sql/group.py:3]: Thanks.

And no, reduce indirections as much as possible.

File: pysparkling/sql/internal_utils/readers/utils.py [hunk 6]
```diff
@@ -1,10 +1,10 @@
-from pysparkling.fileio import File, TextFile
-from pysparkling.sql.casts import get_caster
-from pysparkling.sql.types import (
+from ....fileio import File, TextFile
+from ...casts import get_caster
+from ...types import (
```
[the remark this answers is not in the document]
Comment (svaningelgem) [reply pysparkling/sql/internal_utils/readers/utils.py:3]: I'm a bit on the fence here. Although I would prefer the import as it is now.
isort takes care of doubles if they would appear. Plus, I think the code is more clear when you do not have the `types.` always in front of it...

File: pysparkling/rdd.py (hunk 2 above)
Comment (svaningelgem) [reply pysparkling/rdd.py:170]: ok

**Review by svaningelgem — COMMENTED** [review]
I went through the comments and wrote my train of thought. Can you check if this changes your mind?

File: pysparkling/context.py (hunk 1 above)
Comment (svaningelgem) [reply pysparkling/context.py:196]: A sample of the error I'm getting:
```
pysparkling\__init__.py:8: in <module>
    from . import exceptions, fileio, streaming
pysparkling\streaming\__init__.py:2: in <module>
    from .context import StreamingContext
pysparkling\streaming\context.py:4: in <module>
    from .dstream import DStream
pysparkling\streaming\dstream.py:5: in <module>
    from ..rdd import EmptyRDD
pysparkling\rdd.py:23: in <module>
    from .context import Context
pysparkling\context.py:17: in <module>
    from .rdd import RDD
E   ImportError: cannot import name 'RDD'
```

The problem is that everything is depending on RDD/Row and Context. I tried playing around a bit with the order of the imports, but I failed to find a good fit.

And both rdd & context have the other imported...
As RDD is only needed 2x in context.py, I opted to make it inline there. (Context appears 165x in rdd.py).

File: pysparkling/rdd.py [hunk 7]
```diff
@@ -2080,10 +2080,9 @@ def toDF(self, schema=None, sampleRatio=None):
...
         [Row(age=1, name='Alice')]
         """
-        # Top level import would cause cyclic dependencies
-        # pylint: disable=import-outside-toplevel
-        from pysparkling import Context
-        from pysparkling.sql.session import SparkSession
+        # pylint: disable=import-outside-toplevel  # against a cyclic-import
+        from .sql.session import SparkSession
+
         sparkSession = SparkSession._instantiatedSession or SparkSession(Context())
```
Comment (svenkreiss) [inline pysparkling/rdd.py:2086]: I don't know why a new context is created here. The import issues probably come from here. There is a `self.context` that you get without importing `Context`. If you can change those, can you avoid importing Context at the top of the file?

File: pysparkling/context.py (hunk 1 above)
Comment (svenkreiss) [reply pysparkling/context.py:196]: This was working before. This is a blocker for me. See below in RDD file for a suggestion.

**Review by svenkreiss — CHANGES_REQUESTED** [review]

File: pysparkling/rdd.py (hunk 7 above)
Comment (svaningelgem) [reply pysparkling/rdd.py:2086]: Thanks for your help! I fixed it as such and now it works fine.

File: pysparkling/sql/session.py [hunk 8]
```diff
@@ -261,6 +260,9 @@ def prepare(obj):
             def prepare(obj):
                 return obj
 
+        # pylint: disable=import-outside-toplevel, cyclic-import
+        from ..rdd import RDD
```
Comment (svenkreiss) [inline pysparkling/sql/session.py:264]: My understanding is that the import structure within the `sql` module is complicated and not great at the moment. However, the core classes outside of the `sql` module should be fine. Do you mind re-checking whether these imports outside of top-level are really necessary for those cases like this one?

File: pysparkling/sql/group.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (svenkreiss) [reply pysparkling/sql/group.py]: Don't use `F`. Avoid indirections.

[the remark this answers is not in the document]
Comment (svaningelgem) [reply pysparkling/sql/group.py]:
```
import functions:
E   ModuleNotFoundError: No module named 'functions'

from . import functions:
*** Works fine

import .functions
E     File "E:\pysparkling\pysparkling\sql\group.py", line 1
E       import .functions
E              ^
E   SyntaxError: invalid syntax
```

So I don't think I have any other way than to call it the way I did? I don't know how to reduce the complexity of the imports here.
Could you provide me some pointers?

Thanks

[the remark this answers is not in the document]
Comment (svenkreiss) [reply pysparkling/sql/group.py]: There seems to be some misunderstanding. Use `from . import functions`. You say that works fine. Use that.

[the remark this answers is not in the document]
Comment (svenkreiss) [reply pysparkling/sql/group.py]: The "indirection" to avoid is that someone has to look up what `F` is.

File: pysparkling/sql/session.py (hunk 8 above)
Comment (svaningelgem) [reply pysparkling/sql/session.py]: I was able to move a few, but the complete cleanup of these imports would lead me too far. I'm going to move it into an issue (as it very likely will become) to rationalize this later on.

File: pysparkling/__version__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (svenkreiss) [reply pysparkling/__version__.py:1]: I have to come back to this. In `context.py` this also happened before: Context is imported from `__init__.py` but inside of `context.py` the `__version__` is imported from `__init__.py`. 

I assume because of the order that `__version__` was the first thing defined, this doesn't lead to a cycle. This should still be possible.

File: pysparkling/sql/session.py [hunk 9]
```diff
@@ -1,21 +1,20 @@
...
-import pysparkling
-from pysparkling import RDD
-from pysparkling.context import Context
-from pysparkling.sql.conf import RuntimeConfig
-from pysparkling.sql.dataframe import DataFrame
-from pysparkling.sql.internals import DataFrameInternal
-from pysparkling.sql.readwriter import DataFrameReader
-from pysparkling.sql.schema_utils import infer_schema_from_list
-from pysparkling.sql.types import (
+from .. import Context, RDD
```
Comment (svenkreiss) [inline pysparkling/sql/session.py:3]: In case circular imports are still blocking, this line might be an issue. Use `from ..context import Context` and `from ..rdd import RDD` instead.

**Review by svenkreiss — CHANGES_REQUESTED** [review]

**Comment by svaningelgem** [comment]
The only thing which is blocking now is this circular import:
```
scripts\tcpperf_server.py:1:0: R0401: Cyclic import (pysparkling -> pysparkling.context) (cyclic-import)
```

Which is correct. `__init__` is importing Context, and Context is importing the __version__ and accumulators from .

I now changed a few things and the cyclic imports are gone... However they're appearing again when I'm reverting this particular change:
```
sql\session.py: 
from .. import __version__
```

This is way too fickle to my liking. That's the reason why I introduced the __version__ file so you could load the version directly from the file, and don't need to go through loading the whole module.

Additional benefit in my opinion is that the version is in it's own file, you don't need to remove the `get_versions` method anymore. (= cleaner code)

**Comment by svenkreiss** [comment]
Thank you for this major effort! LGTM.

**Review by svenkreiss — APPROVED** [review]
